src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -206,7 +207,14 @@ caseToScheme values caseAlternatives =
           (concatMapWithIndex
            (\i b -> (binderToTest (vectorRef (cdr value) (IntegerLiteral i)) b))
            binders)
-    binderToTest _ _ = error "Not implemented"
+
+    -- We have a level of indirection. `value' is bound to `binder' through
+    -- `ident'. For instance in
+    --     data Foo = Bar Int
+    --     match bar@(Bar i) = bar
+    -- `value' is an Identifier referring to (Bar i), ident is an Ident "bar"
+    -- and binder is the ConstructorBinder for (Bar i).
+    binderToTest value (NamedBinder _ann _ident binder) = binderToTest value binder
```

Comment:
I am confused about why we throw away identifiers here and in other places.
Like, what happens when we have something like this? (this might be a useful test to include anyways)
```
data Foo = Bar Int

doubleFoo :: Foo -> Int
doubleFoo bar@(Bar i) = case bar of
  Bar j -> i + j
```

Reply:
```
(define Bar
  (lambda (value0)
    (cons (quote Bar) (vector value0))))

(define doubleFoo
  (lambda (v)
    (cond ((eq? (car v) (quote Bar))
	   (cond ((eq? (car v) (quote Bar))
		  (+ (vector-ref (cdr v) 0)
		     (vector-ref (cdr v) 0))))))))
```

It seems to be working but it definitely needs some more work: we should get rid of the nested `cond`s and remove the redundant tests.

Reply:
For a comparison, the generated JS:

```
var Bar = (function () {
    function Bar(value0) {
        this.value0 = value0;
    };
    Bar.create = function (value0) {
        return new Bar(value0);
    };
    return Bar;
})();

var doubleFoo = function (v) {
    return v.value0 + v.value0 | 0;
};
```

Reply:
>I am confused about why we throw away identifiers here and in other places.

The reason here is that `bar` is an alias to `v` (thus the comment regarding the level of indirection). `value` is an `Expr.Var "v"`, and ident is an `Ident.Ident "bar"`. We throw the ident away here because we don't need it to generated the test part of this cond clause. But we use it later in `variablesToReplace` in order to make the substitution `bar` -> `v`.

Reply:
Summing it up, we have a few things going on here and we should create some cards:
- A function matching a single constructor type should not check for the type, to make the behavior on par with the JS compiler. This is issue #27 but maybe it could be better phrased and described.
- Nested conds should be flattened. Not sure if we should do it by emitting better code or by brutally handling it at the optimizer level.
- Redudant cond clauses should be dropped. This could be tricky.

Reply:
> The reason here is that `bar` is an alias to `v` (thus the comment regarding the level of indirection). `value` is an `Expr.Var "v"`, and ident is an `Ident.Ident "bar"`. We throw the ident away here because we don't need it to generated the test part of this cond clause. But we use it later in `variablesToReplace` in order to make the substitution `bar` -> `v`.

Oh I see now, thanks for the clarification - I guess this explanation could go into expanding that comment.

> Summing it up, we have a few things going on here and we should create some cards

I think there should be cards about this, but I'll note that all of this is optimization and thus lower priority than getting all the necessary places in order (e.g. FFI), so I'd say not to worry excessively about this right now